README.md

```diff
@@ -538,6 +538,61 @@ dependencies, making any newly generated `requirements.txt` environment-dependen
 As a general rule, it's advised that users should still always execute `pip-compile`
 on each targeted Python environment to avoid issues.
 
+### Maximizing reproducibility
+
+`pip-tools` is a great tool to improve the reproducibility of builds.
+But there are a few things to keep in mind.
+
+- `pip-compile` will produce different results in different environments as described in the previous section.
+- `pip` must be used with the `PIP_CONSTRAINT` environment variable to lock dependencies in build environments as documented in [#8439](https://github.com/pypa/pip/issues/8439).
+- Dependencies come from many sources.
+
+Continuing the `pyproject.toml` example from earlier creating a single lock file could be done like:
+
+```text
+$ pip-compile --all-build-distributions --all-extras --output-file=constraints.txt --strip-extras pyproject.toml
+#
+# This file is autogenerated by pip-compile with Python 3.9
+# by the following command:
+#
+#    pip-compile --all-build-distributions --all-extras --output-file=constraints.txt --strip-extras pyproject.toml
+#
+asgiref==3.5.2
+    # via django
+attrs==22.1.0
+    # via pytest
+django==4.1
+    # via my-cool-django-app (pyproject.toml)
```

Comment:
By the way, how would we show in these comments if a dependency came from `build-system.requires` or specific dynamically called hooks?

Reply:
@apljungquist I like option 2 the most. Though it would probably need to be documented, right?

Reply:
I'm still skeptical because I feel like it adds complexity, both to the implementation and to the UI, and that few people will benefit from the extra specificity. But I gave it a try so let me know what you think.

Reply:
I think this looks nice in the example output and it's worth it to me.

Reply:
Alright. Can we mark this conversation as resolved?